scrapy/core/downloader/handlers/s3.py

```diff
@@ -24,21 +24,29 @@ def __init__(self, settings, *,
         # If no credentials could be found anywhere,
         # consider this an anonymous connection request by default;
         # unless 'anon' was set explicitly (True/False).
-        anon = kw.get('anon')
-        if anon is None and not aws_access_key_id and not aws_secret_access_key:
-            kw['anon'] = True
-        self.anon = kw.get('anon')
+        self.anon = kw.pop('anon', None)
+        if kw:
+            raise TypeError(f'Unexpected keyword arguments: {kw}')
 
         self._signer = None
         import botocore.auth
-        import botocore.credentials
-        kw.pop('anon', None)
-        if kw:
-            raise TypeError(f'Unexpected keyword arguments: {kw}')
-        if not self.anon:
-            SignerCls = botocore.auth.AUTH_TYPE_MAPS['s3']
-            self._signer = SignerCls(botocore.credentials.Credentials(
-                aws_access_key_id, aws_secret_access_key))
+        from botocore.credentials import Credentials
+        from botocore.session import Session
+        if self.anon is None or self.anon is False:
+            # Cause Session.create_client() is used in pipelines and feed exporter,
+            # following the credential creation logic in Session.create_client()
+            # maximize consistence of AWS auth across the project.
+            if aws_access_key_id is not None and aws_secret_access_key is not None:
+                credentials = Credentials(access_key=aws_access_key_id,
+                                          secret_key=aws_secret_access_key)
+            else:
+                credentials = Session().get_credentials()
+            if credentials is not None:
+                SignerCls = botocore.auth.AUTH_TYPE_MAPS['s3']
+                self._signer = SignerCls(credentials)
+                self.anon = False
+            else:
+                self.anon = True
```

Comment:
This change of `self.anon` was not happening before. Is it needed?

Reply:
`self.anon = True` happens before this PR, but only by checking `aws_access_key_id is not None and aws_secret_access_key is not None`.

The `self.anon = False` here may not be needed. I'm just emphasizing it's set `False` explicitly, to differentiate the `None` value when `anon` param is not passed during `S3DownloadHandler(..., anon=True/False)` init.

Reply:
> `self.anon = True` happens before this PR

Sorry, I had not realize that.

I’m trying to understand the change of behavior compared to the previous implementation, because we will want to at least cover it in the release notes:

- `self.anon` can now become `True` even if the credential parameters and settings are falsy, as long as the credentials can be automatically determined from the environment (by [botocore’s credential resolver](https://github.com/boto/botocore/blob/0bf25028b9fa66f030a72daf8499128394e45ba7/botocore/credentials.py#L56)).
  
  While I think this change of behavior is for the better, it does mean that restoring the previous behavior from settings alone is now impossible. That is, before you could force anonymous access from settings by letting them have falsy values; now that is not possible. I think maybe we should add a setting to allow users to force anonymous credentials, just in case.

- `self.anon` now becomes `False` if credentials are not provided and cannot be found.

  It is not a big change, but since it’s not needed functionality-wise and is a change in behavior that could affect user-defined subclasses, I think it would be best to avoid it and leave `self.anon` as `None` if that was its value.

Reply:
Thought for a while. The logic before this PR onlys set `.anon` `True` if it's `None`. In my implementation, the valued could be changed from `False` to `True. It does breaks the anonymous function. I'll try to fix.

Reply:
> That is, before you could force anonymous access from settings by letting them have **falsy** values; now that is not possible. I think maybe we should add a setting to allow users to force anonymous credentials, just in case.

You misunderstood the logic here. Anonymous is declared with `True`, not falsy values.

If a user declares anonymous, the s3 request will be ensured anonymous, in both previous and current implementation. The PR here doesn't break this.

---

The problem is, what if a user declares non-anonymous with `.anon = False` explcitly and can't get a credential? We haven't take this branch into consideration before this PR.

Keyword `anon` was firstly introduced in 1b4fd3a by nyov. The reason behind it is still written in the comment.

```
        # If no credentials could be found anywhere,
        # consider this an anonymous connection request by default;
        # unless 'anon' was set explicitly (True/False).
```

I broke the "unless 'anon' was set explicitly (True/False)" part. We should change the code to raise an exc if no credential could be found but `.anon == False`.

Reply:
> That is, before you could force anonymous access from settings by letting them have **falsy** values; now that is not possible. I think maybe we should add a setting to allow users to force anonymous credentials, just in case.

You misunderstood the logic here. Anonymous is declared with `True`, not falsy values.

If a user declares anonymous, the s3 request will be ensured anonymous, in both previous and current implementation. The PR here doesn't break this.

---

The problem is, what if a user declares non-anonymous with `.anon = False` explcitly and can't get a credential? We haven't take this branch into consideration before this PR.

Keyword `anon` was firstly introduced in 1b4fd3a by nyov. The reason behind it is still written in the comment.

```
        # If no credentials could be found anywhere,
        # consider this an anonymous connection request by default;
        # unless 'anon' was set explicitly (True/False).
```

I broke the "unless 'anon' was set explicitly (True/False)" part. We should change the code to raise an exc if no credential could be found but `.anon == False`.

Reply:
> We should change the code to raise an exc if no credential could be found but `.anon == False`.

That makes a lot of sense, and even though raising an exception here is also a backward-incompatible change, I also believe this one is for the better as well.

> > That is, before you could force anonymous access from settings by letting them have **falsy** values; now that is not possible. I think maybe we should add a setting to allow users to force anonymous credentials, just in case.
> 
> You misunderstood the logic here. Anonymous is declared with `True`, not falsy values.

I think I actually failed to properly explain my point. I meant falsy values for the `aws_access_key_id` and `aws_secret_access_key` parameters and their corresponding settings, `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`.

We need to keep in mind that users usually interact with this class through settings, not through `__init__` parameters. So forget about `__init__` parameters, and think of user interactions with this class solely from settings.

Before, they could get `self.anon` to be set to `True` by simply not using any of those 2 settings:

```python
        if not aws_access_key_id:
            aws_access_key_id = settings['AWS_ACCESS_KEY_ID']
        if not aws_secret_access_key:
            aws_secret_access_key = settings['AWS_SECRET_ACCESS_KEY']
        # …
        if anon is None and not aws_access_key_id and not aws_secret_access_key:
            kw['anon'] = True
```

Now, in an environment where there _are_ credentials, there’s no way to get `self.anon` to be set to `True` from settings alone. Even if those two settings are undefined, if the environment has credentials, `self.anon` will never be `True`, environment credentials will always be used.

That’s why I think we might want to expose that as a setting now, so that users that need to prevent environment credentials to be picked can do so through settings. Maybe this could be enough:

```python
self.anon = kw.pop('anon', settings['AWS_ANONYMOUS_ACCESS'])
```

Although if the `anon` parameter is `None`, maybe it should get overridden by the setting, so maybe this would be best:

```python
self.anon = kw.pop('anon', None)
if self.anon is None:
    self.anon = settings['AWS_ANONYMOUS_ACCESS']
```

Reply:
> But `AWS_ANONYMOUS_ACCESS` sounds like we're enabling anonymous for all aws related things. Could we define it more specific, e.g. `DOWNLOAD_HANDLERS_S3_ANONYMOUSE`?

To follow a similar pattern to those used in Scrapy to solve similar situations, what we can do is, in addition to defining the `AWS_ANONYMOUS_ACCESS` setting (which may be used to get that behavior as a default for anything using S3 in Scrapy), allowing to pass S3 credential settings on a per-request basis through `Request.meta`. The download handler can then use the meta parameters to override the global parameters when processing a request.

> I propose to remove `kw` argument (deprecate it first?)

Given that the pre-existing code already raises an exception if it has anything other than `anon`, I think we can replace it with an `anon` keyword-only parameter. It will raise a different exception, but I imagine that’s fine.

As for deprecating first, I was going to say that it may not be necessary in this case, but it’s probably best to be extra careful and do that, log a deprecation error before raising `ValueError` for now, and eventually remove the `**kw` parameter for good.

Reply:
> which may be used to get that behavior as a default for anything using S3 in Scrapy

aws s3 is used in S3DownloadHandler for downloading, in pipeline and feed exporter for uploading. Setting up s3 without any credential for downloading means the resource is publicly accessible/readable. But for uploading, I can't imagine s3 bucket is setup publicly writable for all users. It's a security fault.

"Instance metadata service" is an alternative for this publicly writable need. It provides a credential (without any settings) but only could be got from specific Amazon EC2 instance. 


> allowing to pass S3 credential settings on a per-request basis through `Request.meta`

This is another feature, we'd better open a separate PR for it.

> Given that the pre-existing code already raises an exception if it has anything other than anon, I think we can replace it with an anon keyword-only parameter.

Thanks. It makes sense.

Reply:
> Still, do we have an application scenario for this feature: disable credential usage for a specific s3 request? The purpose of the intro of `anon` is to skip reading credential info those `~/.aws/credentials`, `~/.aws/config` files.

I think it’s possible for a spider to aim to download input data from a private S3 storage (e.g. source URLs), but force anonymous access to public S3 URLs that are crawled. I’m not sure if publicly readable  S3 storages fail on unexpected authentication, but we definitely do not want to leak our S3 credentials to untrusted S3 servers. Specially since S3-compatible services other than Amazon’s exist.

So, I do see a scenario where per-request customization of credentials makes sense. That said, it is not part of the scope of this pull request, to get this merged we only need to make it possible to make S3 downloads anonymous.

Since there is a clear download/upload distinction, as you mentioned, I think your idea may be best to keep the changes minimal: implement just a setting but name it so that it’s clear that it does not affect other S3 credential usages. Seeing other Scrapy settings that alter download handlers, I suggest a name like `DOWNLOADER_S3_ANONYMOUS`.